node/src/components/block_synchronizer/global_state_synchronizer.rs

```diff
@@ -243,19 +242,34 @@ impl GlobalStateSynchronizer {
         let to_fetch = self.fetch_queue.take(num_fetches_to_start);
 
         for (trie_hash, request_root_hashes) in to_fetch {
-            let peers = request_root_hashes
+            let peers: Vec<_> = request_root_hashes
                 .iter()
                 .filter_map(|request_root_hash| self.request_states.get(request_root_hash))
                 .flat_map(|request_state| request_state.peers.iter().copied())
                 .collect();
-            effects.extend(effect_builder.fetch_trie(trie_hash, peers).event(
-                move |trie_accumulator_result| Event::FetchedTrie {
-                    trie_hash,
-                    request_root_hashes,
-                    trie_accumulator_result,
-                },
-            ));
-            self.in_flight.insert(trie_hash);
+            if peers.is_empty() {
+                // if we have no peers, requeue the trie - trie accumulator would return an error,
+                // anyway
+                debug!(%trie_hash, "no peers available for requesting trie; requeuing");
```

Comment:
That's a good question. I think, with the current state of the code, the answer is "no". It could be notified of new peers with a new request for the same root hash, but AFAIK we don't re-issue requests from the block synchronizer if the peer set changes.

Should we just fail here, then?

Reply:
So we either need the block synchronizer to notify the global state synchronizer of new peers (here Ed would prefer the global state synchronizer to be able to use the block builder's peer list somehow - but that would require major modifications to the code), or just fail and make sure that the block synchronizer retries on a failure. The latter should be simpler for the time being.

Reply:
Sounds good to me! (Also, doesn't have to be fixed in this PR necessarily.)

Reply:
I'll make it fail at the very least, or we might get some trie nodes stuck in the queue forever and causing some spam in the logs :sweat_smile: